packages/block-editor/src/components/simulate-media-query/index.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * External dependencies
+ */
+import { filter, get } from 'lodash';
+import { match } from 'css-mediaquery';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect } from '@wordpress/element';
+
+const ENABLED_MEDIA_QUERY = '(min-width:0px)';
+const DISABLED_MEDIA_QUERY = '(min-width:999999px)';
+
+const VALID_MEDIA_QUERY_REGEX = /\((min|max)-width:[^\(]*?\)/g;
+
+function getStyleSheetsThatMatchHostname() {
+	return filter(
+		get( window, [ 'document', 'styleSheets' ], [] ),
+		( styleSheet ) => {
+			return (
+				styleSheet.href &&
+				styleSheet.href.includes( window.location.hostname )
+			);
+		}
+	);
+}
+
+function isReplaceableMediaRule( rule ) {
+	if ( ! rule.media ) {
+		return false;
+	}
+	// Need to use "media.mediaText" instead of "conditionText" for IE support.
+	return !! rule.media.mediaText.match( VALID_MEDIA_QUERY_REGEX );
+}
+
+function replaceRule( styleSheet, newRuleText, index ) {
+	styleSheet.deleteRule( index );
+	styleSheet.insertRule( newRuleText, index );
+}
+
+function replaceMediaQueryWithWidthEvaluation( ruleText, widthValue ) {
+	return ruleText.replace( VALID_MEDIA_QUERY_REGEX, ( matchedSubstring ) => {
+		if (
+			match( matchedSubstring, {
+				type: 'screen',
+				width: widthValue,
+			} )
+		) {
+			return ENABLED_MEDIA_QUERY;
+		}
+		return DISABLED_MEDIA_QUERY;
+	} );
+}
+
+/**
+ * Function that manipulates media queries from stylesheets to simulate a given viewport width.
+ *
+ * @param {Array} partialPaths Paths of stylesheets to manipulate.
+ * @param {number} width Viewport width to simulate.
+ */
+export default function useSimulatedMediaQuery( width ) {
```

Comment:
How can we test this properly? unit tests? e2e-tests?

Reply:
Hmm that's a good question. 

Unit testing this function is tricky, because the only thing it does is cause DOM-related side-effects. We'd need to mock fetching a stylesheet,  build a little "stylesheet" fixture for it to go through, and we'd probably have to reimplement the `deleteRule` and `insertRule` methods from CSSStyleSheet (I don't think `jsdom` has this API). Then we'd need to run it inside a dummy component and check what happens on mount and unmount.

On the other hand, e2e tests would have to rely on specific styles that change across breakpoints. So we'd be checking e.g if `randomElement.style.fontSize === '16px'` at a certain viewport size, and minor styling changes could easily break them. 

All things considered, I'm more inclined towards e2e in this case because they'll test the actual functionality, as opposed to the implementation details. 

We could still  write unit tests for the smaller functions such as `isReplaceableMediaRule` and `replaceMediaQueryWithWidthEvaluation` though.

Reply:
#20159 created to follow up on this.